### PR TITLE
Single Search Result

### PR DESCRIPTION
**Pull request by megandouli**
Changed searcher to return 1 result from the database (not Elasticsearch) if query contains subject and course number.

### CODE REVIEW

File: backend/searcher.ts [hunk 1]
```diff
@@ -256,24 +284,35 @@ class Searcher {
...
+    } else {
+      const searchResults = await this.getSearchResults(query, termId, min, max, filters);
+      ({ resultCount, took, aggregations } = searchResults);
+      const startHydrate = Date.now();
+      results = await (new HydrateSerializer()).bulkSerialize(searchResults.output);
+      hydrateDuration = Date.now() - startHydrate;
+    }
 
-    return {
+    const res = {
```
Comment (gamburgm) [inline backend/searcher.ts]: Mind just returning `res`?

File: backend/searcher.ts [hunk 2]
```diff
@@ -256,24 +284,35 @@ class Searcher {
...
-    // this can be re-written in a way that's less bad
-    const {
-      output, resultCount, took, aggregations,
-    } = await this.getSearchResults(query, termId, min, max, filters);
+    let results; let resultCount; let hydrateDuration; let took; let aggregations;
 
-    const startHydrate = Date.now();
-    const results = await (new HydrateSerializer()).bulkSerialize(output);
+    // if we know that the query is of the format of a course code, we want to return only one result
+    const courseCodePattern: RegExp = /^\s*([a-zA-Z]{2,4})\s*(\d{4})?\s*$/i;
```
Comment (gamburgm) [inline backend/searcher.ts]: Might make sense to move this elsewhere, e.g. to the constructor, and make it all caps or something--treat it like an OOD constant.

File: backend/searcher.ts [hunk 3]
```diff
@@ -256,24 +284,35 @@ class Searcher {
...
+    const patternResults = query.match(courseCodePattern);
+    const subject = patternResults ? patternResults[1].toUpperCase() : '';
+    if (patternResults && typeof Number(patternResults[2]) === 'number' && (this.getSubjects()).has(subject)) {
```
Comment (gamburgm) [inline backend/searcher.ts]: This function is getting difficult to understand--would be worth sitting down and revisiting this, seeing how we might make the code easier to understand.

**Review by gamburgm — CHANGES_REQUESTED** [review]
Looks good overall, made a few small comments :^)

File: backend/searcher.ts [hunk 4]
```diff
@@ -246,6 +250,30 @@ class Searcher {
...
+    return {
+      results: results,
+      resultCount: result === null ? 0 : 1,
+      took: 0,
+      hydrateDuration: Date.now() - start,
+      aggregations: result ? this.getSingleResultAggs(result) : { nupath: [], subject: [], classType: [] },
+    };
+  }
+
+  getSingleResultAggs(result: CourseWithSections) : AggResults {
```
Comment (NEUDitao) [inline backend/searcher.ts:261]: can't we just return this directly w/o the `const agg`

**Review by NEUDitao — APPROVED** [review]

**Pull request merged**